experimental/parser/lexer.go

```diff
@@ -0,0 +1,675 @@
+		if err == nil {
+			token.SetValue(tok, value)
+			return tok
+		}
+
+		if err.(*strconv.NumError).Err == strconv.ErrRange {
+			l.Error(ErrIntegerOverflow{Token: tok})
+		} else {
+			l.Error(ErrInvalidNumber{Token: tok})
+		}
+	case strings.HasPrefix(digits, "0b"): // Binary ints are an extension.
+		value, err := strconv.ParseUint(strings.TrimPrefix(digits, "0b"), 2, 64)
+		if err == nil {
+			token.SetValue(tok, value)
+			return tok
+		}
+
+		if err.(*strconv.NumError).Err == strconv.ErrRange {
+			l.Error(ErrIntegerOverflow{Token: tok})
+		} else {
+			l.Error(ErrInvalidNumber{Token: tok})
+		}
+	}
+
+	// This is either a float or a decimal number. Try parsing as decimal first, otherwise
+	// try parsing as float.
+	_, err := strconv.ParseUint(digits, 10, 64)
+	if err == nil {
+		// This is the most common result. It gets computed ON DEMAND by token.Token.AsNumber.
+		// DO NOT place it into l.literals.
+		return tok
+	}
+
+	// Check if it's out of range, because if that's the only problem, it's
+	// guaranteed to parse as a possibly-infinite float.
+	if err.(*strconv.NumError).Err == strconv.ErrRange {
+		// We want this to overflow to Infinity as needed, which ParseFloat
+		// will do for us. Otherwise it will ties-to-even as expected. Currently,
+		// the spec does not say "ties-to-even", but it says "nearest value",
+		// which everyone says when they mean "ties-to-even" whether they know it
+		// or not.
+		//
+		// ParseFloat itself says it "returns the nearest floating-point number
+		// rounded using IEEE754 unbiased rounding", which is just a weird way to
+		// say "ties-to-even".
+		value, _ := strconv.ParseFloat(digits, 64)
+		token.SetValue(tok, value)
+		return tok
+	}
+
+	// If it was not well-formed, it might be a float.
+	value, err := strconv.ParseFloat(digits, 64)
+
+	// This time, the syntax might be invalid. If it is, we decide whether
+	// this is a bad float or a bad integer based on whether it contains
+	// periods.
+	if err.(*strconv.NumError).Err == strconv.ErrSyntax {
+		l.Error(ErrInvalidNumber{Token: tok})
+	}
+
+	// If the error is ErrRange we don't care, it will clamp to Infinity in
+	// the way we want, as noted by the comment above.
+	token.SetValue(tok, value)
+	return tok
+}
+
+// LexString lexes a string starting at the current cursor.
+//
+// The cursor position should be just before the string's first quote character.
+func (l *lexer) LexString() token.Token {
+	start := l.cursor
+	q := l.Pop()
+
+	// Seek to the end of the string, unescaping as we go. We do not
+	// materialize an unescaped string if this string does not require escaping.
+	var buf strings.Builder
+	var haveEsc bool
+escapeLoop:
+	for !l.Done() {
+		r := l.Pop()
+		if r == q {
+			break
+		}
+
+		// Warn if the user has a non-printable character in their string that isn't
+		// ASCII whitespace.
+		if !unicode.IsGraphic(r) && !strings.ContainsRune(" \n\t\r", r) {
```

Comment:
FWIW, it's not possible for `\n` (unescaped) to be part of a string literal: protobuf doesn't support multi-line string literals ([spec](https://protobuf.com/docs/language-spec#string-literals)). That would instead be an unterminated string literal.

Reply:
I'm gonna leave it here for future-proofing, though.